### DISCUSSION TITLE
DNS Settings

### DISCUSSION THREAD

**Discussion by stellarpower** [gazzasat/huawei-cpe#2] [Ideas]
Hi,

Thanks for this! I got a router second-hand, locked to Three. Hadn't realised how much they had modded the firmware. But nice to see a number of people in the FOSS world have been early adopters of 5G and trying to push back and open things up.

I would look at implementing and make an MR but I'm currently trying to diagnose intermittent connection issues, so not the best time to be trying to push code. I'm not the only one finding duclkduckgo, cppreference and github seems to be partiularly difficult to connect to - worst nightmare. So I'll just going pin the link here as an idea:

https://www.youtube.com/watch?v=fkn0PqA8GPk&ab_channel=MarcBehar

I have no DNS settings available in my firmware build, and it seems even in stock they have located them unintuitively in the DHCP section. The one-line snippet `$('#dhcp_dns').show();` worked for me though. FYI this was to switch to Cloudflare's DNS. I have heard some mention of their WARP pseudo-VPN as being a good solution, not yet tested it. For reasons that aren't clear to me I've heard that simply pointing to their DNS can help in some situations of dropping connections. And if nothing else it's meant to be faster than the network provider's own nameservers probably are.

I've also been unsuccessful thus far at enabling the APN settings dialogue box so if I ever have time I might try to commit some support for that too. Simply removing the hide CSS class didn't do it and I was just using F12, not a proper script as you have.

Cheers

**Comment by gingerbeardman** [comment]
DNS is already exposed in this script. 

After you activate it you see three buttons top left: 
1. BANDS
2. INFO
3. DNS

Clicking DNS takes you to that page, where you can change to whatever you want and hit SAVE.

https://github.com/gazzasat/huawei-cpe/blob/33e7f1a68afbc959da7e9bea7d4e2ab7027db45f/src/huawei-bands-util.js#L4-L7

https://github.com/gazzasat/huawei-cpe/blob/33e7f1a68afbc959da7e9bea7d4e2ab7027db45f/src/huawei-bands-util.js#L195

**Comment by gazzasat** [comment]
As @gingerbeardman mentions the DNS setting works well and enables you to override the providers DNS. I have been using the Quad 9 DNS servers for a while and they have been fast and trouble free.

A while back I tried unsuccessfully to access the APN settings dialog on the Three branded unlocked routers.  With some script you can enable the dialog and text boxes but the API call to SAVE the settings has been hard disabled / removed in the firmware as far as I can see.

Any Three routers that are 'unlocked' should be OK with using other provider SIMs where the required APN is the primary one as the router will auto-select the first / default APN for that SIM - or if the provider for the SIM ignores the APN (like Smarty).  

In the UK I successfully tested with Three ( including Smarty ), EE and Vodafone but some other MVNOs like TalkTalk did not work as the default APN was some other PAYG voda one.  Basically the unit won't register and connect to the provider network if the APN is not right.

I guess you'll have to give other SIMs a test to be sure.

**Comment by stellarpower** [comment]
Ah yes, apologies :sweat_smile:. Don't know how I missed that. Besides regular connection failures to github the latest firefox update also broke the developer console so I hadn't run the script in several days.

On the APN front I popped in another SIM and word on the grapevine is their modded firmware hard sets the "unlock remaining" count to zero so there is almost nothing to be done. Given they make such a fuss they sell their phones unlocked it's disappointing. Mine isn't even one of the 3-branded once. I would assume trying to reflash would be unsuccessful, even if I could get hold of the unmodified firmware. Wouldn't mind being able to see the APN settings at least, when I "dis-disable" them I usually don't see anything helpful on the screen, just broken UI controls. I have the URL for the endpoint though, which returns everything as XML and that works, so may give it a try again.
